wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/tool/kubernetes/KubernetesEnvironmentProvisioner.java

```diff
@@ -0,0 +1,167 @@
+import org.eclipse.che.api.devfile.server.exception.DevfileException;
+import org.eclipse.che.api.devfile.server.exception.DevfileFormatException;
+import org.eclipse.che.api.workspace.server.model.impl.EnvironmentImpl;
+import org.eclipse.che.api.workspace.server.model.impl.MachineConfigImpl;
+import org.eclipse.che.api.workspace.server.model.impl.RecipeImpl;
+import org.eclipse.che.api.workspace.server.model.impl.WorkspaceConfigImpl;
+import org.eclipse.che.commons.lang.Pair;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesRecipeParser;
+import org.eclipse.che.workspace.infrastructure.openshift.environment.OpenShiftEnvironment;
+
+/** @author Sergii Leshchenko */
+public class KubernetesEnvironmentProvisioner {
+  @VisibleForTesting static final String YAML_CONTENT_TYPE = "application/x-yaml";
+
+  private final KubernetesRecipeParser objectsParser;
+
+  @Inject
+  public KubernetesEnvironmentProvisioner(KubernetesRecipeParser objectsParser) {
+    this.objectsParser = objectsParser;
+  }
+
+  /**
+   * Provision default environment with the specified kubernetes/openshift tool.
+   *
+   * <p>If there is already a default environment with kubernetes/openshift then content will be
+   * updated with new list.
+   *
+   * @param workspaceConfig workspace where recipe should be provisioned
+   * @param environmentType type of environment that should be provisioned. Should be {@link
+   *     KubernetesEnvironment#TYPE} or {@link OpenShiftEnvironment#TYPE}
+   * @param toolObjects parsed objects of the specified tool
+   * @throws DevfileRecipeFormatException if exception occurred during existing environment parsing
+   * @throws DevfileRecipeFormatException if exception occurred during kubernetes object
+   *     serialization
+   * @throws DevfileException if any other exception occurred
+   */
+  public void provision(
+      WorkspaceConfigImpl workspaceConfig,
+      String environmentType,
+      List<HasMetadata> toolObjects,
+      Map<String, MachineConfigImpl> machines)
+      throws DevfileException, DevfileRecipeFormatException {
+    String defaultEnv = workspaceConfig.getDefaultEnv();
+    EnvironmentImpl environment = workspaceConfig.getEnvironments().get(defaultEnv);
+    if (environment == null) {
+      checkItemsHasUniqueKindToName(toolObjects);
+
+      RecipeImpl recipe =
+          new RecipeImpl(environmentType, YAML_CONTENT_TYPE, asYaml(toolObjects), null);
+      String envName = "default";
+      EnvironmentImpl env = new EnvironmentImpl(recipe, emptyMap());
+      env.getMachines().putAll(machines);
+      workspaceConfig.getEnvironments().put(envName, env);
+      workspaceConfig.setDefaultEnv(envName);
+    } else {
+      RecipeImpl envRecipe = environment.getRecipe();
+
+      environment.getMachines().putAll(machines);
```

Comment:
It should not occur since machine are provision only by `DockerimageToolToWorkspaceApplier` and machine name is a unique name of tool, but it makes sense to add a check to avoid further hide bug. Will add.

Reply:
Added a check.